### ISSUE TITLE
Two valid copies of the same dataset don't necessarily contain the same audio

### ISSUE THREAD

**Issue by lostanlen** [soundata/soundata#51]
Consider the following script:

```python
import soundata

urbansound8k = soundata.initialize('urbansound8k')
urbansound8k.download()  # download the dataset
val = urbansound8k.validate()  # validate that all the expected files are there
assert val == ({'metadata': {}, 'clips': {}}, {'metadata': {}, 'clips': {}})

example_clip = urbansound8k.clip(urbansound8k.clip_ids[0])

jam1 = example_clip.to_jams()
jam2 = example_clip.to_jams()
print(jam1==jam2)
```

For reasons of experimental reproducibility, i would expect the above script to
(1) execute without error (assuming the dataset validates)
(2) print `True` on every execution of the script
(3) print `True` on every machine

Unfortunately, none of these three useful properties are guaranteed. Because the `data_home` folder (here, implicitly passed to `Dataset` via the default kwarg `data_home=None`) may be writable for other threads, the result of `jam1==jam2` is nondeterministic. We cannot predict it before `jam2` is computed. Worse, the `jam1` assignment may pass while the `jam2` assignment may fail. 

All these nondeterministic effects may happen despite running `urbansound8k.validate()` before the creation of the `example_clip`. In fact, they could potentially happen even if we ran `urbansound8k.validate()` again after creating `jam1`.

The problem is not limited to `urbansound8k` but concerns all currently available modules:
 * `esc50`
 * `tau2019uas`
 * `tut2017se`
 * `urbansed`
 * `urbansound8k`
 
 I wonder if there's a plan to ensure that `clip.to_jams()` returns constant values across calls?
 
Besides the `to_jams()` converter, i am under the impression that the access to audio is not reliable either. 
 
 ```
import numpy as np
import soundata

def get_first_audio(dataset_name)
        dataset = soundata.initialize(dataset_name)
        dataset.download()  # download the dataset
        val = dataset.validate()  # validate that all the expected files are there
        assert val == ({'metadata': {}, 'clips': {}}, {'metadata': {}, 'clips': {}})
        
        example_clip = dataset.clip(dataset.clip_ids[0])
        return example_clip.audio

x, sr_x = get_first_audio('urbansound8k')
y, sr_y = get_first_audio('urbansound8k')
print(np.all(x==y) and (sr_x==sr_y))
```

Because the two calls are asynchronous, the script above may occasionally print `False` on some machines. Like in the `jam1==jam2` example, the first call to `get_first_audio` may pass while the second may fail. And again, calling `validate` repeatedly on the `clip` before accessing the audio would not guarantee that `np.all(x==y) and (sr_x==sr_y)`.

How can i be sure that two valid copies of the same dataset of my machine contain the same audio?

Lastly, two successive calls to `soundata.initialize(dataset_name).validate()` for the same `dataset_name` will not necessarily return the same tuple:

```python
val1 = soundata.initialize('urbansound8k').validate()
val2 = soundata.initialize('urbansound8k').validate()
assert (val1 == val2)
```

The assertion above is not always `True`. How much should i trust the output of `validate` if it isn't consistent across sucessive calls? How often do you recommend that i call `validate` in a typical research workflow?

**Comment by justinsalamon** [comment]
Thanks for the detailed report @lostanlen !

Let's start with the jams converter - IIUC the issue is that if you were to execute the example script you have provided in concurrent threads, one thread may overwrite the data in the download folder already written by another thread, such that the audio path changes, and since the path is documented in the jams object, jams1 and jams2 end up not being equal. Is this correct?

If this is indeed the issue, it would be helpful to understand when/why you would execute the download function concurrently in multiple threads? Our assumption is that downloading a dataset is something you only want to do once, after which you work with the same copy of the data - you can still run validation at the beginning of every experimental script to ensure the data haven't been modified/invalidated. 

But maybe I'm missing something/misunderstanding. Perhaps you can provide a concrete use case where this issue arises? This would be helpful in getting further understanding of the issue. IIUC, the audio loading issue would also stem from the concurrent execution of the `download` function, which is not something we designed for, so I'm hoping that by understanding/addressing the jams issue we can take care of both birds.

Thanks!

**Comment by lostanlen** [comment]
> Let's start with the jams converter - IIUC the issue is that if you were to execute the example script you have provided in concurrent threads, one thread may overwrite the data in the download folder already written by another thread, such that the audio path changes, and since the path is documented in the jams object, jams1 and jams2 end up not being equal. Is this correct?

You're right. The most likely multi-threaded scenario i imagine would be to run soundata in interactive mode:
```
jam1 = example_clip.to_jams()
# at this point i minimize the Python terminal window,
# switch to a different application (Finder, audio editor, text editor ...),
# accidentally edit some audio or annotations in ~/sound_datasets,
# and eventually return to my Python session.
jam2 = example_clip.to_jams()
print(jam1==jam2)
```
The `print` statement above may return `False` even though the `example_clip` object is unchanged.
Note that the part within comments (`# switch to a different application ... `) may in fact result of the action of a different user of the computer (or high-performance computing facility) having write access to shared memory. So even if i'd keep staring at the `example_clip` object, it could return different `to_jams()` at different times due to the action of another user on the `data_home`.

The problem may also arise in a single-threaded scenario. Idioms of the form:
```
jam1 = example_clip.to_jams()
opaque_function()
jam2 = example_clip.to_jams()
print(jam1==jam2)
```
may print `True` or `False` depending on the content of `opaque_function`. An example of such `opaque_function` would be a time stretching effect that would overwrite the audio of the `example_clip` and change its duration.

Here i am giving the `to_jams` function as an example but the same problem happens with accessing `.audio` (except `.audio`, unlike `to_jams()`, is cached so the situation is a bit more complicated. Thus i chose `to_jams()` for simplicity)


> If this is indeed the issue, it would be helpful to understand when/why you would execute the download function concurrently in multiple threads?

The issue has nothing to do with `download`. The `download` function could be called once, multiple times, or never at all, and the problem would still arise. The reason why i included `download` in my example script was to follow strictly soundata's `README.md` — and also to signify that the problem may arise on freshly downloaded datasets. In other words, this is not a contrived example.

Maybe my issue title is not good, though. I wrote "Two valid copies of the same dataset don't necessarily contain the same audio", but the problem may arise even with a single copy as long as that copy is mutable on disk (which is always the case: nothing is strictly read-only in the filesystem). I'm happy to edit the title to better reflect the context of the issue.

What i'm trying to say is that this is not the kind of problem that can be solved by writing unit tests, nor by re-running the code on a controlled environment. Indeed, if we're absolutely sure that nothing happens on disk between the two `to_jams()` callsite, then we can be confident that `jam1==jam2` will be `True`. In a sense, this makes the problem even harder to debug.

To be clear, i am not concerned about the picture-perfect, "textbook" uses of soundata, but the real-world messy ones. To err is human, and i would like to imagine soundata as a safety belt for at least certain kinds of human error in experimentation.

> If this is indeed the issue, it would be helpful to understand when/why you would execute the download function concurrently in multiple threads? Our assumption is that downloading a dataset is something you only want to do once

That's right, i have the same assumption. I would not redownload a dataset unless `soundata` tells me explicitly that there are missing files in my copy.


> after which you work with the same copy of the data - you can still run validation at the beginning of every experimental script to ensure the data haven't been modified/invalidated.

That's my whole point: running validation at the beginning of every experimental script does not ensure that the data aren't modified _during_ the experimental script.
To put it bluntly: how useful is a smoke detector that only "validates" the absence of smoke when i press a button, and remains idle afterwards?


> Perhaps you can provide a concrete use case where this issue arises? 

My use case is, i suppose, the same as everyone's: i am clumsy and impatient.

Clumsy: sometimes i listen to audio files in a dataset. Sometimes i copy them. Sometimes i `rsync` them to a different machine. Sometimes `rsync` breaks when 90% done and i don't notice. Sometimes i apply data augmentation and accidentally overwrite the original files. Sometimes a colleague hands me a shiny new `opaque_function` which i run without knowing what it does. Sometimes i run `validate` and don't interpret the result properly. Sometimes my data processing pipelines involve more than one programming language. Sometimes i do all of these things at once, on the same computer, in the rush towards a deadline.

Impatient: I like to run Python in interactive mode. I like to benchmark models by running many SLURM jobs on the same dataset with many different hyperparameters. I want to benchmark my models on as many datasets as possible, thus i download a lot. I am happy to re-run `validate` on a regular basis (say, once a week or so) but i don't like to have to set myself a reminder for that specifically.  I don't like to have to redownload things unless absolutely necessary. I like to access the `.audio` property more than once and assume that it will always return the same value, no matter when i access it for the first time. And, out of all things, i hate having to re-do my benchmark from the start because i didn't notice an annotation mismatch three months ago.

Does that make sense?


One last thing: i don't want to sound too negative. I truly appreciate the effort for reproducibility of tools like `soundata` and i believe that it can be a game-changer for rapid adoption of new datasets and for more consistent benchmarking of machine listening systems. But at the same time, i think it's important to acknowledge the risk of an "illusion of security" on the part of users. Just because there is a `validate` function and i take the time to run it after downloading the dataset, or even at the start of every session, doesn't mean that the audio and annotation will still be valid when i use it later in the same session.

 
 > Thanks!
 
I hope this helps :)

**Comment by justinsalamon** [comment]
Thanks @lostanlen , that definitely helps me understand the scenario.

It sounds to me (and please correct me if i'm wrong) that there is a tradeoff here: the safest option would be to always validate-on-load all assets whether they be audio, jams, or other forms of metadata. The down-side is that this would significantly slow down execution as we'd have to recompute the md5 hash of every file whenever it's used, to the point where for some users/use-cases it could render the library unusable.

What if we added a `validate_on_load` flag to every function that loads data from disk (be it audio, jams, or otherwise), which would be set to `False` by default but could be set to `True` by the user? This way we would support on-the-fly validation, but not force it on users by default. I think this could be a good compromise?

**Comment by justinsalamon** [comment]
p.s.

> One last thing: i don't want to sound too negative. I truly appreciate the effort for reproducibility of tools like soundata and i believe that it can be a game-changer for rapid adoption of new datasets and for more consistent benchmarking of machine listening systems. But at the same time, i think it's important to acknowledge the risk of an "illusion of security" on the part of users. Just because there is a validate function and i take the time to run it after downloading the dataset, or even at the start of every session, doesn't mean that the audio and annotation will still be valid when i use it later in the same session.

I never had any doubt that your only goal is to help make `soundata` better 😃🙏

**Comment by lostanlen** [comment]
yes, i think that a validate-on-load procedure would be great. It can take the form of a kwarg during `initialize` or it could be a custom `with` statement, like so:

```python
with dataset.open(clip_id) as clip:
    # here, soundata validates the clip automatically.
    # if file does not exist -> `FileNotFoundError`
    # if we don't have read access -> `PermissionError`
    # if checksum is wrong -> `IOError`
    
    # at this point we can safely call librosa
    y, sr = clip.audio
    jam = clip.to_jams()
    
# here we can safely access y, sr, and jam variables
```

In this way, users only pay for what they use. If they only use a fraction of the dataset, it's not necessary to validate unused clips.

> It sounds to me (and please correct me if i'm wrong) that there is a tradeoff here: the safest option would be to always validate-on-load all assets whether they be audio, jams, or other forms of metadata

You're right. Validating every time would definitely be overkill. 

A way to mitigate that tradeoff would be to cache the result of `validate` with `joblib.Memory`. The cache lifespan could be user-defined, e.g. 1 day or 1 week by default. In this way, the validation overhead would be amortized across multiple runs, and people could navigate the speed/safety tradeoff by changing the cache lifespan constant.

To give an example on how it would work, @beasteers once wrote a PR for librosa which proposes to reset the cache every 24 hours.
https://github.com/librosa/librosa/pull/1142

The case `validate_on_load=False` would mean that the cache has an infinite lifespan and is never reset: it would only be updated if the user calls `dataset.validate()` manually.

P.S.: if `with` statements seem verbose to you, we could also write functions like `dataset.librosa_load(clip_id, *args, **kwargs)` which would do the cache lookup (and/or validation) automatically. This mechanic could be implemented generically as part of a custom `@validated` decorator. But maybe i'm drawing conclusions too far here.

**Comment by lostanlen** [comment]
With all due respect, @magdalenafuentes, i think that this is more than an `enhancement` issue.

The `0.1.0` version of soundata, announced today, advertises that 

> Soundata makes it easy to [...] validate that a downloaded dataset is complete and perfectly matches a canonical version 

I think that the claim above is exaggerated, for the reasons i listed above. I'm afraid that this might give people a feeling of false security: i.e., interpret the advert above as "validate once and worry no longer".

Yet, as of v0.1.0, running `val = urbansound8k.validate()` does not guarantee that two successive calls to `initialize` will return the same Python object. In fact, even after the construction of the `dataset` object, it does not even guarantee that `dataset.clip(clip_id)` returns the same Python object.

Quoting the sentence above, the risk that a "downloaded dataset" is "complete and perfectly matches a canonical version" is fairly small considering that we implement downloaders in terms of DOI's, which are immutable by design. However, the risk of overwriting a local copy of the dataset is far higher, especially in the age of shared/distributed memory volumes. 

I am concerned that leaving the return value of expressions like `dataset.clip(clip_id).to_jams()` to be nondeterministic opens the door to an array of inscrutable bug reports, because it not only depends upon the value of `dataset` and  `clip_id`, but also on the entire state of the local machine and its shared volumes. Since the user will never be able to reproduce this state (let alone post it on GitHub), a large number of bugs will remain unsolved—to say nothings of the bugs which won't even be reported because they'll go unnoticed.

My current understanding of the `validate` function is that it's like a smoke alarm button which you need to press everytime you return home to "validate" that your house hasn't caught fire. Hence my question: what's the recommended use of `validate` in the current state of soundata? How often should i run it to "validate that a downloaded dataset is complete and perfectly matches a canonical version"?

**Comment by magdalenafuentes** [comment]
Hey @lostanlen, thanks for bringing up this issue again. I think you're right, and sorry if it looked like we were undervaluing the problem. We have discussed recently with @rabitt what would be the best way to address this and what is exactly that we want and can do about it.

My understanding is that the main issue right now is that `validate` is simply a tool to check data consistency at the dataset level and users would run it either after the download is complete or at the beginning of the work pipeline. That's better than what was there before (not checking at all) but still doesn't cover the case where e.g. you're working with some model and while running it in a dataset you might change the data without even noticing, and you would assume that data is right because you validated at the beginning of your work. Am I getting it right? 

One first iteration of a solution is to check files when loading in memory (clip-wise instead of at the dataset level), but then for things that are cached that would not work? But at least we could include that tool for the user and clarify exactly in what cases the validation would be consistent and when it would not? Thoughts?
